d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -313,20 +313,17 @@ XdsClient.D2SymlinkNodeResourceWatcher getSymlinkResourceWatcher(String symlinkN
       @Override
       public void onChanged(String resourceName, XdsClient.D2SymlinkNodeUpdate update)
       {
-        // update maps between symlink name and actual node name, listen to the actual node
+        // Update maps between symlink name and actual node name
         String actualResourceName = update.getNodeData().getMasterClusterNodePath();
-        if (resourceName.contains(D2_CLUSTER_NODE_PREFIX))
-        {
-          String actualNodeName = removeNodePathPrefix(actualResourceName, D2_CLUSTER_NODE_PREFIX);
-          updateSymlinkAndActualNodeMap(symlinkName, actualNodeName);
-          listenToCluster(actualNodeName);
-        }
-        else
-        {
-          String actualNodeName = removeNodePathPrefix(actualResourceName, D2_URI_NODE_PREFIX);
-          updateSymlinkAndActualNodeMap(symlinkName, actualNodeName);
-          listenToUris(actualNodeName);
-        }
+        String actualNodeName = getNodeName(actualResourceName);
+        updateSymlinkAndActualNodeMap(symlinkName, actualNodeName);
+        // listen to the actual nodes
+        // Note: since cluster symlink and uri parent symlink always point to the same actual node name, and it's a
+        // redundancy and a burden for the symlink-update tool to maintain two symlinks for the same actual node name,
+        // we optimize here to use the cluster symlink to listen to the actual nodes for both cluster
+        // and uri parent.
+        listenToCluster(actualNodeName);
+        listenToUris(actualNodeName);
```

Comment:
Just a thought, would we also want to delete the existing symlink properties from the property bus, and let the requests fail in case the `listenToUris(actualNodeName);` or cluster calls fail? Because if we don't remove those properties and fetching new properties from xDS fails due to some reason, traffic would still keep on flowing to the old leader (~master~), which is not one might want.

Reply:
that's true. But I'd prefer to keep parity with zk flow (which will also point to the old leader) for now to ease INDIS migration monitoring. We could create a backlog item to improve this after the migration.

Reply:
Makes sense.